File: binding/ios/EagleAppTest/PerformanceTest/PerformanceTest.swift
//
//  Copyright 2023 Picovoice Inc.
//  You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
//  file accompanying this source.
//  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
//  specific language governing permissions and limitations under the License.
//

import Foundation
import XCTest

import Eagle

class PerformanceTest: XCTestCase {
    let accessKey: String = "{TESTING_ACCESS_KEY_HERE}"
    let iterationString: String = "{NUM_TEST_ITERATIONS}"
    let indexThresholdString: String = "{INDEX_PERFORMANCE_THRESHOLD_SEC}"
    let searchThresholdString: String = "{SEARCH_PERFORMANCE_THRESHOLD_SEC}"

    override func setUp() {
        super.setUp()
        continueAfterFailure = false
    }

    func testProfilerPerformance() throws {
        try XCTSkipIf(indexThresholdString == "{ENROLL_PERFORMANCE_THRESHOLD_SEC}")

        let numTestIterations = Int(iterationString) ?? 30
        let indexPerformanceThresholdSec = Double(indexThresholdString)
        try XCTSkipIf(indexPerformanceThresholdSec == nil)

        let bundle = Bundle(for: type(of: self))
        let fileURL: URL = bundle.url(forResource: "test", withExtension: "wav")!
        let audioData = try Data(contentsOf: fileURL)
        var pcm = [Int16](repeating: 0, count: (audioData.count - 44) / 2)
        _ = pcm.withUnsafeMutableBytes {
            audioData.copyBytes(to: $0, from: 44..<audioData.count)
        }

        let eagleProfiler = try EagleProfiler(accessKey: accessKey)

        var results: [Double] = []
        for _ in 0...numTestIterations {
            var totalNSec = 0.0

            let before = CFAbsoluteTimeGetCurrent()
            (_, _) = try eagleProfiler.enroll(pcm: pcm)
            let after = CFAbsoluteTimeGetCurrent()
            totalNSec += (after - before)
            results.append(totalNSec)
        }
        eagleProfiler.delete()

        let avgNSec = results.reduce(0.0, +) / Double(numTestIterations)
        let avgSec = Double(round(avgNSec * 1000) / 1000)
        XCTAssertLessThanOrEqual(avgSec, indexPerformanceThresholdSec!)
    }

    func testProcPerformance() throws {
        try XCTSkipIf(searchThresholdString == "{PROC_PERFORMANCE_THRESHOLD_SEC}")

        let numTestIterations = Int(iterationString) ?? 30
        let searchPerformanceThresholdSec = Double(searchThresholdString)
        try XCTSkipIf(searchPerformanceThresholdSec == nil)

        let bundle = Bundle(for: type(of: self))

        let enrollUrls: [URL] = [
            bundle.url(forResource: "speaker_1_utt_1", withExtension: "wav")!,
            bundle.url(forResource: "speaker_1_utt_2", withExtension: "wav")!
        ]
        let eagleProfiler = try EagleProfiler(accessKey: accessKey)
        for enrollUrl in enrollUrls {
            let audioData = try Data(contentsOf: enrollUrl)
            var pcm = [Int16](repeating: 0, count: (audioData.count - 44) / 2)
            _ = pcm.withUnsafeMutableBytes {
                audioData.copyBytes(to: $0, from: 44..<audioData.count)
            }

            (_, _) = try eagleProfiler.enroll(pcm: pcm)
        }
        let profile = try eagleProfiler.export()
        eagleProfiler.delete()

        let testAudioURL: URL = bundle.url(forResource: "test", withExtension: "wav")!
        let audioData = try Data(contentsOf: testAudioURL)
        var pcm = [Int16](repeating: 0, count: (audioData.count - 44) / 2)
        _ = pcm.withUnsafeMutableBytes {
            audioData.copyBytes(to: $0, from: 44..<audioData.count)
        }

        let eagle = try Eagle(accessKey: accessKey, speakerProfiles: [profile])

        var results: [Double] = []
        for _ in 0...numTestIterations {
            var totalNSec = 0.0

            let before = CFAbsoluteTimeGetCurrent()
            _ = try eagle.process(pcm: pcm)
            let after = CFAbsoluteTimeGetCurrent()
            totalNSec += (after - before)
            results.append(totalNSec)
        }
        eagle.delete()

        let avgNSec = results.reduce(0.0, +) / Double(numTestIterations)
        let avgSec = Double(round(avgNSec * 1000) / 1000)
        XCTAssertLessThanOrEqual(avgSec, searchPerformanceThresholdSec!)
    }
}

File: binding/android/EagleTestApp/eagle-test-app/src/androidTest/java/ai/picovoice/eagle/testapp/PerformanceTest.java
/*
    Copyright 2023 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.eagle.testapp;

import static org.junit.Assert.assertTrue;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Assume;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;

import ai.picovoice.eagle.Eagle;
import ai.picovoice.eagle.EagleProfile;
import ai.picovoice.eagle.EagleProfiler;

@RunWith(AndroidJUnit4.class)
public class PerformanceTest extends BaseTest {

    int numTestIterations = 30;

    @Before
    public void Setup() throws IOException {
        super.Setup();
        String iterationString = appContext.getString(R.string.numTestIterations);

        try {
            numTestIterations = Integer.parseInt(iterationString);
        } catch (NumberFormatException ignored) { }
    }

    @Test
    public void testProfilerPerformance() throws Exception {
        String enrollThresholdString = appContext.getString(R.string.enrollPerformanceThresholdSec);
        Assume.assumeNotNull(enrollThresholdString);
        Assume.assumeFalse(enrollThresholdString.equals(""));

        double enrollPerformanceThresholdSec = Double.parseDouble(enrollThresholdString);

        EagleProfiler eagleProfiler = new EagleProfiler.Builder()
                .setAccessKey(accessKey)
                .build(appContext);

        File audioFile = new File(testResourcesPath, testPath);
        short[] pcm = readAudioFile(audioFile.getAbsolutePath());

        long totalNSec = 0;
        for (int i = 0; i < numTestIterations + 1; i++) {
            long before = System.nanoTime();
            eagleProfiler.enroll(pcm);
            long after = System.nanoTime();

            // throw away first run to account for cold start
            if (i > 0) {
                totalNSec += (after - before);
            }
        }

        double avgNSec = totalNSec / (double) numTestIterations;
        double avgSec = ((double) Math.round(avgNSec * 1e-6)) / 1000.0;
        assertTrue(
                String.format(
                    "Expected threshold (%.3fs), profiler took (%.3fs)",
                    enrollPerformanceThresholdSec,
                    avgSec),
                avgSec <= enrollPerformanceThresholdSec
        );

        eagleProfiler.delete();
    }

    @Test
    public void testProcPerformance() throws Exception {
        String procThresholdString = appContext.getString(R.string.procPerformanceThresholdSec);
        Assume.assumeNotNull(procThresholdString);
        Assume.assumeFalse(procThresholdString.equals(""));

        double procPerformanceThresholdSec = Double.parseDouble(procThresholdString);

        EagleProfiler eagleProfiler = new EagleProfiler.Builder()
                .setAccessKey(accessKey)
                .build(appContext);

        for (String path : enrollPaths) {
            File audioFile = new File(testResourcesPath, path);
            short[] pcm = readAudioFile(audioFile.getAbsolutePath());
            eagleProfiler.enroll(pcm);
        }

        EagleProfile profile = eagleProfiler.export();

        Eagle eagle = new Eagle.Builder()
                .setAccessKey(accessKey)
                .setSpeakerProfile(profile)
                .build(appContext);

        File audioFile = new File(testResourcesPath, testPath);
        short[] pcm = readAudioFile(audioFile.getAbsolutePath());
        int numFrames = pcm.length / eagle.getFrameLength();

        long totalNSec = 0;
        for (int i = 0; i < numTestIterations + 1; i++) {
            long before = System.nanoTime();
            for (int j = 0; j < numFrames; j++) {
                eagle.process(Arrays.copyOfRange(
                        pcm,
                        i * eagle.getFrameLength(), (i + 1) * eagle.getFrameLength())
                );
            }
            long after = System.nanoTime();

            // throw away first run to account for cold start
            if (i > 0) {
                totalNSec += (after - before);
            }
        }

        double avgNSec = totalNSec / (double) numTestIterations;
        double avgSec = ((double) Math.round(avgNSec * 1e-6)) / 1000.0;
        assertTrue(
                String.format("Expected threshold (%.3fs), process took (%.3fs)", procPerformanceThresholdSec, avgSec),
                avgSec <= procPerformanceThresholdSec
        );
    }
}
